modules/apps/29-fee/types/msgs.go

```diff
@@ -41,3 +44,135 @@ func (msg MsgRegisterCounterpartyAddress) GetSigners() []sdk.AccAddress {
+		Relayers:        relayers,
+	}
+}
+
+// ValidateBasic performs a basic check of the MsgPayPacketFee fields
+func (msg MsgPayPacketFee) ValidateBasic() error {
+	// validate channelId
+	err := host.ChannelIdentifierValidator(msg.SourceChannelId)
+	if err != nil {
+		return err
+	}
+
+	// validate portId
+	err = host.PortIdentifierValidator(msg.SourcePortId)
+	if err != nil {
+		return err
+	}
+
+	// signer check
+	_, err = sdk.AccAddressFromBech32(msg.Signer)
+	if err != nil {
+		return sdkerrors.Wrap(err, "failed to convert msg.Signer into sdk.AccAddress")
+	}
+
+	// enforce relayer is nil
+	if msg.Relayers != nil {
+		return ErrRelayersNotNil
+	}
+
+	// ensure fee's are valid amounts
+	if !msg.Fee.AckFee.IsValid() {
+		return sdkerrors.ErrInvalidCoins
+	}
+
+	if !msg.Fee.ReceiveFee.IsValid() {
+		return sdkerrors.ErrInvalidCoins
+	}
+
+	if !msg.Fee.TimeoutFee.IsValid() {
+		return sdkerrors.ErrInvalidCoins
+	}
+
+	return nil
+}
+
+// GetSigners implements sdk.Msg
+func (msg MsgPayPacketFee) GetSigners() []sdk.AccAddress {
+	signer, err := sdk.AccAddressFromBech32(msg.Signer)
+	if err != nil {
+		panic(err)
+	}
+	return []sdk.AccAddress{signer}
+}
+
+// NewMsgPayPacketAsync creates a new instance of MsgPayPacketFee
+func NewMsgPayPacketFeeAsync(identifiedPacketFee IdentifiedPacketFee, signer string) *MsgPayPacketFeeAsync {
+	return &MsgPayPacketFeeAsync{
+		IdentifiedPacketFee: identifiedPacketFee,
+		Signer:              signer,
+	}
+}
+
+// ValidateBasic performs a basic check of the MsgPayPacketFeeAsync fields
+func (msg MsgPayPacketFeeAsync) ValidateBasic() error {
+	// validate channelId
+	err := host.ChannelIdentifierValidator(msg.IdentifiedPacketFee.PacketId.ChannelId)
+	if err != nil {
+		return err
+	}
+
+	// validate portId
+	err = host.PortIdentifierValidator(msg.IdentifiedPacketFee.PacketId.PortId)
+	if err != nil {
+		return err
+	}
+
+	// signer check
+	_, err = sdk.AccAddressFromBech32(msg.Signer)
+	if err != nil {
+		return sdkerrors.Wrap(err, "failed to convert msg.Signer into sdk.AccAddress")
+	}
+
+	// enforce relayer is nil
+	if msg.IdentifiedPacketFee.Relayers != nil {
+		return ErrRelayersNotNil
+	}
+
+	// ensure sequence is not 0
+	if msg.IdentifiedPacketFee.PacketId.Sequence == 0 {
+		return sdkerrors.ErrInvalidSequence
+	}
+
+	// ensure fee's are valid amounts
+	if !msg.IdentifiedPacketFee.Fee.AckFee.IsValid() {
+		return sdkerrors.ErrInvalidCoins
+	}
+
+	if !msg.IdentifiedPacketFee.Fee.ReceiveFee.IsValid() {
+		return sdkerrors.ErrInvalidCoins
+	}
+
+	if !msg.IdentifiedPacketFee.Fee.TimeoutFee.IsValid() {
+		return sdkerrors.ErrInvalidCoins
+	}
+
+	return nil
+}
+
+// GetSigners implements sdk.Msg
+func (msg MsgPayPacketFeeAsync) GetSigners() []sdk.AccAddress {
+	signer, err := sdk.AccAddressFromBech32(msg.Signer)
+	if err != nil {
+		panic(err)
+	}
+	return []sdk.AccAddress{signer}
+}
+
+func NewIdentifiedPacketFee(packetId *channeltypes.PacketId, fee Fee, relayers []string) *IdentifiedPacketFee {
```

Comment:
maybe we should add a validate basic check for identified packet fee (and then do `msg.IdentifiedPacketFee.ValidateBasic()` in async msg

We could also add a ValidateBasic to Fee which checks that the coins are valid and at least one of the coins is non empty

Reply:
could be done in a follow if we think this is desirable